rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,261 @@
+            return DEFAULT_DECODING_OUTPUT
+
+        return self._decode_swap(
+            context=context,
+            fee_raw=fee_raw,
+            in_asset=buy_asset,
+            in_amount=buy_amount,
+        )
+
+    def _decode_swap_token_to_native(self, context: DecoderContext) -> DecodingOutput:
+        tx_log = context.tx_log
+        sender = context.transaction.from_address
+        native_token = self.evm_inquirer.native_token
+        if bytes_to_address(tx_log.topics[2]) == self.router_address:
+            in_event = None
+            fee_raw = None
+            for event in context.decoded_events:
+                # FCind the receive event
+                if (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.RECEIVE
+                    event.counterparty = CPT_RAINBOW_SWAPS
+                    event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+                    event.address = self.router_address
+                    in_event = event
+
+                    # Find the log withdraw the weth token
+                    for log in context.all_logs:
+                        if log.topics[0] == WETH_WITHDRAW_TOPIC:
+                            try:
+                                in_amount = asset_raw_value(in_event.amount, native_token)
+                            except ConversionError:
+                                in_amount = 0
+                            # The fee will be the withdraw data subtract the in_amount
+                            fee_raw = int.from_bytes(log.data) - in_amount
+
+            if not (fee_raw and in_event):
+                return DEFAULT_DECODING_OUTPUT
+
+            sell_amount_raw = int.from_bytes(tx_log.data)
+            sell_asset = self.base.get_or_create_evm_token(tx_log.address)
+            amount = asset_normalized_value(sell_amount_raw, sell_asset)
+            tx_hash = context.transaction.tx_hash
+
+            out_event = self.base.make_event(
+                tx_hash=tx_hash,
+                sequence_index=in_event.sequence_index,
+                timestamp=context.transaction.timestamp,
+                event_type=HistoryEventType.TRADE,
+                event_subtype=HistoryEventSubType.SPEND,
+                asset=sell_asset,
+                amount=amount,
+                location_label=sender,
+                notes=f'Swap {amount} {sell_asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap',  # noqa: E501
+                counterparty=CPT_RAINBOW_SWAPS,
+                address=self.router_address,
+            )
+            # Need to change the index of receive to respect the events order
+            in_event.sequence_index += 1
```

Comment:
instead of doing this check the maybe_reshuffle, not sure if it is what you need but just in case. Also we have a function `get_next_sequence_index` that you can use. This one looks chaotic and error prone